### PR TITLE
TT-7584 Configure OpenTelemetry tracing backend

### PR DESCRIPTION
**Pull request by tbuchaillot**
## Description

Added OpenTelemetry configuration and init on gateway start

## Related Issue

https://tyktech.atlassian.net/browse/TT-7584
## Motivation and Context

https://tyktech.atlassian.net/browse/TT-7584
## How This Has Been Tested

Added unit tests
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit:  
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5403545436)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit:  
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5403545436)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 3df205a6b78c868badc1f891d612ba99ed97696c 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5405868131)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: b3cf73a4d73a0f5849c14a2fb560e4cacfc10117 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5409517888)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 61e9a4fa4f3ed88a099102a87df2233bbbafd11b 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5410416800)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 210f703573d4adc1a89887ae05d47cda3d1de595 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5411306406)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 7826658b46354fa766a9cd68e4bbfb3280c635db 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5411763476)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 007ffaa097948a68beae033290930936c12968f2 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5413986487)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 6d3e0e7aa78fe55c2a6120cb29f51b00b3f892d0 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5413980591)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: b88c7a845a00c1377a78bb507814f8ca509dd15d 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5420727911)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 2cae2d80691b99e93d05870a0f6a3b64f6b73b66 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5424224454)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 3609aec2c7f67018888b14dce5a4b57c43c75374 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5447290985)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: aaef06c47777988b15fc2599fe12eae48dd97971 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5447321538)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 6ee1f2da303a5bc47d0a18bab5ad123ba0890e51 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5447351026)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 4c710041aecf3720228b5a848f731cee1898c483 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5448874856)

File: config/config.go [hunk 1]
```diff
@@ -937,6 +938,9 @@ type Config struct {
 	// Section for configuring OpenTracing support
 	Tracer Tracer `json:"tracing"`
 
+	// Section for configuring Opentelemetry
+	OpenTelemetry otelconfig.OpenTelemetry `json:"opentelemetry"`
```
Comment (sredxny) [inline config/config.go:942]: am wondering how will we publish the documntation of this config option with the docs generator 🤔

File: gateway/otel.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	tyktrace "github.com/TykTechnologies/opentelemetry/trace"
+)
+
+func (gw *Gateway) InitOpenTelemetry() {
+	gwConfig := gw.GetConfig()
+
+	traceLogger := mainLog.WithFields(logrus.Fields{
+		"exporter":           gwConfig.OpenTelemetry.Exporter,
+		"endpoint":           gwConfig.OpenTelemetry.Endpoint,
+		"connection_timeout": gwConfig.OpenTelemetry.ConnectionTimeout,
```
Comment (sredxny) [inline gateway/otel.go:15]: I think we dont really need the timeout printed in each log?

File: gateway/otel_test.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+			gw.SetConfig(tc.givenConfig)
+			gw.afterConfSetup()
+
+			gw.InitOpenTelemetry()
+			assert.NotNil(t, gw.TraceProvider)
+
+			if tc.givenConfig.OpenTelemetry.Endpoint == "invalid" {
+				// making sure that noop tracer provider is used when exporter is invalid
+				assert.Equal(t, oteltrace.NewNoopTracerProvider(), gw.TraceProvider)
```
Comment (sredxny) [inline gateway/otel_test.go:109]: could the noop tracer affect in some way the response time of the gateway? even when it does nothing?

File: gateway/server.go [hunk 4]
```diff
@@ -1389,6 +1391,14 @@ func (gw *Gateway) afterConfSetup() {
 		}
 	}
 
+	if conf.OpenTelemetry.Enabled {
+		conf.OpenTelemetry.SetDefaults()
+
+		if conf.OpenTelemetry.ResourceName == "" {
+			conf.OpenTelemetry.ResourceName = "tyk-gateway"
```
Comment (sredxny) [inline gateway/server.go:1398]: should we move "tyk-gateway" to a const?

**Review by sredxny — CHANGES_REQUESTED** [review]

File: config/config.go (hunk 1 above)
Comment (tbuchaillot) [reply config/config.go:942]: It should work the same? it's already included in the schema generation

File: gateway/otel.go (hunk 2 above)
Comment (tbuchaillot) [reply gateway/otel.go:15]: yes, acceptance criteria of the ticket

File: gateway/otel_test.go (hunk 3 above)
Comment (tbuchaillot) [reply gateway/otel_test.go:109]: nope - that's the idea of no operation tracer

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 765b0749614fb736851eb567bc4c0ed52ff1a29a 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5452149046)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 346aa73c4d26b70f0fcfaf5d7312a9350e100114 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5452129149)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 91036527c26cdd992bab7bae75b6849bcd2903ab 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5452200520)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 8a9e90cf010a0d600bed1b0e10079a50a8831c7b 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5452322658)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 6021ee39424201770286add02e3d521139351abe 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5452762617)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 66bf5dba3b221631027438b562fba57b9d53d360 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5453291557)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 60199525d88744e9e42f69bbd84db0654ede97d4 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5453402325)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 0b06ea8871d1f7662f855373bfbe79d304dec151 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5453491817)

**Review by sredxny — APPROVED** [review]
🚀

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5232/merge
Commit: 096f37cfa3fb8c12d0352b660cddcb7044228db8 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5454479033)

**Pull request merged**